consensus/istanbul/backend/engine.go

```diff
@@ -395,39 +394,71 @@ func (sb *Backend) Prepare(chain consensus.ChainReader, header *types.Header) er
 	delay := time.Unix(header.Time.Int64(), 0).Sub(now())
 	time.Sleep(delay)
 
+	logger := sb.logger.New("func", "Backend.Prepare()", "number", number)
 	// modify the block header to include all the ParentCommits
 	// only do this for blocks which start with block 1 as a parent
 	if number > 1 {
-		// copy over the seals we have saved the previous block
+
+		// In some cases, "Prepare" may be called before sb.core has moved to the next sequence,
+		// preventing signature aggregation.
+		// This typically happens in round > 0, since round 0 typically hits the "time.Sleep()"
+		// above.
+		// When this happens, loop until sb.core moves to the next sequence, with a limit of 500ms.
+		waitForSequenceChange := func() *big.Int {
+			timeout := time.After(500 * time.Millisecond)
+			ticker := time.NewTicker(10 * time.Millisecond)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
+					seq := sb.core.Sequence()
+					if seq != nil && seq.Cmp(header.Number) == 0 {
+						logger.Trace("Current sequence matches header", "cur_seq", seq)
+						return seq
+					}
+				case <-timeout:
+					// TODO(asa): Why is this logged by full nodes?
+					log.Trace("Timed out while waiting for core to sequence change, unable to combine commit messages with ParentAggregatedSeal", "cur_seq", sb.core.Sequence())
+					return nil
+				}
+			}
+		}
+		seq := waitForSequenceChange()
+
 		parentExtra, err := types.ExtractIstanbulExtra(parent)
 		if err != nil {
 			return err
 		}
 		parentAggregatedSeal := parentExtra.AggregatedSeal
-
-		additionalParentSeals := sb.core.ParentCommits()
-		if additionalParentSeals != nil && additionalParentSeals.Size() != 0 {
-			logger.Trace("Combining additional seals with the parent aggregated seal", "additionalSeals", additionalParentSeals.String(), "num", number, "parentAggregatedSeal", parentAggregatedSeal.String())
-			// if we had any seals gossiped to us, proceed to add them to the
-			// already aggregated signature
-			if unionAggregatedSeal, err := istanbulCore.UnionOfSeals(parentExtra.AggregatedSeal, additionalParentSeals); err != nil {
-				logger.Error("Failed to create union of parent aggregated seal", "err", err)
-			} else {
-				// need to pass the previous block from the parent to get the parent's validators
-				// (otherwise we'd be getting the validators for the current block)
-				parentValidators := sb.getValidators(parent.Number.Uint64()-1, parent.ParentHash)
-				// only update to use the union if we indeed provided a valid aggregate signature for this block
-				if err := sb.verifyAggregatedSeal(parent.Hash(), parentValidators, unionAggregatedSeal); err != nil {
-					logger.Error("Failed to combine additional seals with parent aggregated seal.", "err", err)
+		logger = logger.New("parentAggregatedSeal", parentAggregatedSeal.String())
+		logger.Trace("Got ParentAggregatedSeal")
+		if seq != nil && seq.Cmp(header.Number) == 0 {
```

Comment:
It seems to me, that if we couldn't get a seq (`== nil`) or if it's no the seq we want... then there's something wrong and we should return an error.

Reply:
It won't strictly stop us from proposing using the 2f+1 signatures from the block in our chain, so I think we should continue to try to make a block